dask/array/reshape.py

```diff
@@ -98,12 +102,114 @@ def reshape_rechunk(inshape, outshape, inchunks):
 
             result_outchunks[oleft] = tuple(c // cs for c in result_inchunks[ii])
 
+            max_in_chunk = _cal_max_chunk_size(inchunks, ii, ii)
+            result_outchunks = _smooth_chunks(oleft, oi, max_in_chunk, result_outchunks)
+            # Build cross product of result_outchunks[oleft:oi+1]
+            result_inchunks[ii] = _calc_lower_dimension_chunks(
+                result_outchunks, oleft, oi
+            )
             oi = oleft - 1
             ii -= 1
 
     return tuple(result_inchunks), tuple(result_outchunks)
 
 
+def _calc_lower_dimension_chunks(chunks, start, stop):
+    # We need the lower dimension chunks to match what the higher dimension chunks
+    # can be combined to, i.e. multiply the different dimensions
+    return tuple(
+        map(
+            lambda x: reduce(mul, x),
+            product(*chunks[start : stop + 1]),
+        )
+    )
+
+
+def _smooth_chunks(ileft, ii, max_in_chunk, result_inchunks):
+    # The previous step squashed the whole dimension into a single
+    # chunk for ileft + 1 (and potentially combined too many elements
+    # into a single chunk for ileft as well). We split up the single
+    # chunk into multiple chunks to match the max_in_chunk to keep
+    # chunksizes consistent:
+    # ((1, 1), (200)) -> ((1, 1), (20, ) * 10) for max_in_chunk = 20
+    # It's important to ensure that all dimensions before the dimension
+    # we adjust have to have all-1 chunks to respect C contiguous arrays
+    # during the reshaping
+    # Example:
+    # Assume arr = da.from_array(np.arange(0, 6).reshape(2, 3), chunks=(2, 3))
+    # Reshaping to arr.reshape(-1, ) will return
+    # [ 0  1  2  3  4  5  6  7  8  9 10 11]
+    # The first dimension of the reshaped dimension are the chunks with length 2
+    # Assume we split the second dimension into (2, 1), i.e. setting the chunks to
+    # ((2, 2), (2, 1)) and the output chunks to ((4, 2, 4, 2), )
+    # [ 0  1  3  4  2  5  6  7  9 10  8 11]
```

Comment:
```suggestion
    # Then, the result will be different because we first reshape the individual, non-contiguous chunks before concatenating them:
    # [ 0  1  3  4  2  5  6  7  9 10  8 11]
```